File: su/2026-07/06_pgr_과제진행하기.java
import java.util.*;

class Solution {
    static class Task {
        String name;
        int start;
        int left;

        Task(String name, int start, int left) {
            this.name = name;
            this.start = start;
            this.left = left;
        }
    }

    public String[] solution(String[][] plans) {
        List<Task> tasks = new ArrayList<>();
        for (String[] plan : plans) {
            tasks.add(new Task(plan[0], toMin(plan[1]), Integer.parseInt(plan[2])));
        }

        tasks.sort(Comparator.comparingInt(t -> t.start));

        Stack<Task> stack = new Stack<>();
        List<String> answer = new ArrayList<>();

        for (int i = 0; i < tasks.size() - 1; i++) {
            Task cur = tasks.get(i);
            Task next = tasks.get(i + 1);

            int available = next.start - cur.start;

            if (cur.left <= available) {
                answer.add(cur.name);
                available -= cur.left;

                while (available > 0 && !stack.isEmpty()) {
                    Task prev = stack.pop();
                    if (prev.left <= available) {
                        answer.add(prev.name);
                        available -= prev.left;
                    } else {
                        prev.left -= available;
                        stack.push(prev);
                        available = 0;
                    }
                }
            } else {
                cur.left -= available;
                stack.push(cur);
            }
        }

        answer.add(tasks.get(tasks.size() - 1).name);
        while (!stack.isEmpty()) {
            answer.add(stack.pop().name);
        }

        return answer.toArray(new String[0]);
    }

    private int toMin(String time) {
        String[] t = time.split(":");
        return Integer.parseInt(t[0]) * 60 + Integer.parseInt(t[1]);
    }
}

File: su/2026-07/06_pgr_과제진행하기.py
def get_minutes(start):
    h, m = map(int, start.split(":"))
    return h * 60 + m

def solution(plans):
    answer = []
    stack = []
    n = len(plans)

    plans.sort(key=lambda x: (x[1]))

    t = get_minutes(plans[0][1])  # 현재 시간
    stack.append((plans[0][0], int(plans[0][2])))  # 이름, 남은 시간
    i = 1

    while i < n:
        name, st, pt = plans[i]
        nxt_st = get_minutes(st)
        nxt_pt = int(pt)

        while stack and t < nxt_st:  # 붕 뜬 시간 동안 가능한 과제 처리
            gap = nxt_st - t 
            temp_name, temp_rest = stack.pop()

            if gap < temp_rest:  # 과제 처리하고도 과제 못 마친 경우
                stack.append((temp_name, temp_rest - gap))   
                break
            else:  # 이전 과제 완벽히 끝난 경우
                t += temp_rest
                answer.append(temp_name)

        t = nxt_st
        stack.append((name, nxt_pt))
        i += 1

    while stack:
        answer.append(stack.pop()[0])

    return answer

print(solution([["korean", "11:40", "30"], ["english", "12:10", "20"], ["math", "12:30", "40"]]))
print(solution([["science", "12:40", "50"], ["music", "12:20", "40"], ["history", "14:00", "30"], ["computer", "12:30", "100"]]))
print(solution([["aaa", "12:00", "20"], ["bbb", "12:10", "30"], ["ccc", "12:40", "10"]]))